File: mParticle-Apple-SDK/SceneDelegateHandler.m
#import "SceneDelegateHandler.h"
#import "MPILogger.h"
#import "mParticle.h"
@import mParticle_Apple_SDK_Swift;

@interface SceneDelegateHandler ()
@property (nonatomic, strong) id<OpenURLHandlerProtocol> appNotificationHandler;
@end

@implementation SceneDelegateHandler

- (instancetype)initWithAppNotificationHandler:(id<OpenURLHandlerProtocol>)appNotificationHandler {
    self = [super init];
    if (self) {
        _appNotificationHandler = appNotificationHandler;
    }
    return self;
}

// Non-extractable: MPILogDebug is an ObjC macro that captures the call site and reads
// MParticle's log level, so Swift composes the messages and this emits them unchanged.
- (void)logLines:(NSArray<NSString *> *)lines {
    for (NSString *line in lines) {
        MPILogDebug(@"%@", line);
    }
}

#if TARGET_OS_IOS
- (void)handleURLContext:(UIOpenURLContext *)urlContext API_AVAILABLE(ios(13.0)) {
    NSString *eventAttribution = nil;
    if (@available(iOS 14.5, *)) {
        eventAttribution = [NSString stringWithFormat:@"%@", urlContext.options.eventAttribution];
    }

    [self logLines:[MPSceneDelegateLogic urlContextLogLinesWithURL:[NSString stringWithFormat:@"%@", urlContext.URL]
                                                 sourceApplication:urlContext.options.sourceApplication
                                                        annotation:[NSString stringWithFormat:@"%@", urlContext.options.annotation]
                                                  eventAttribution:eventAttribution
                                                       openInPlace:urlContext.options.openInPlace]];

    NSDictionary<NSString *, id> *options = [MPSceneDelegateLogic openURLOptionsWithSourceApplication:urlContext.options.sourceApplication];

    [self.appNotificationHandler openURL:urlContext.URL options:options];
}
#endif

- (void)handleUserActivity:(NSUserActivity *)userActivity {
    NSString *userInfoDescription = userActivity.userInfo
        ? [NSString stringWithFormat:@"%@", userActivity.userInfo]
        : [NSString stringWithFormat:@"%@", @{}];

    [self logLines:[MPSceneDelegateLogic userActivityLogLinesWithActivityType:userActivity.activityType
                                                                        title:userActivity.title
                                                          userInfoDescription:userInfoDescription
                                                                   webpageURL:userActivity.webpageURL.absoluteString]];

    (void)[self.appNotificationHandler continueUserActivity:userActivity
                                        restorationHandler:^(__unused NSArray<id<UIUserActivityRestoring>> * _Nullable restorableObjects) {
                                        }];
}

@end

File: mParticle-Apple-SDK/SceneDelegateHandler.swift
import Foundation

@objc
public protocol OpenURLHandlerProtocol {
    func open(_ url: URL, options: [String: Any]?)
    func continueUserActivity(
        _ userActivity: NSUserActivity,
        restorationHandler: @escaping ([UIUserActivityRestoring]?) -> Void
    ) -> Bool
}

@objcMembers
public class SceneDelegateHandler: NSObject {
    private let logger: MPLog
    private let appNotificationHandler: OpenURLHandlerProtocol
    
    public init(logger: MPLog, appNotificationHandler: OpenURLHandlerProtocol) {
        self.logger = logger
        self.appNotificationHandler = appNotificationHandler
    }
    
    #if os(iOS)
    @available(iOS 13.0, *)
    @available(iOSApplicationExtension 13.0, *)
    public func handle(urlContext: UIOpenURLContext) {
        logger.debug("Opening URLContext URL: \(urlContext.url)")
        logger.debug("Source: \(String(describing: urlContext.options.sourceApplication ?? "unknown"))")
        logger.debug("Annotation: \(String(describing: urlContext.options.annotation))")
        if #available(iOS 14.5, *) {
            logger.debug("Event Attribution: \(String(describing: urlContext.options.eventAttribution))")
        }
        logger.debug("Open in place: \(urlContext.options.openInPlace ? "True" : "False")")

        let options = ["UIApplicationOpenURLOptionsSourceApplicationKey": urlContext.options.sourceApplication];
        
        self.appNotificationHandler.open(urlContext.url, options: options as [String: Any])
    }
    #endif
    
    public func handleUserActivity(_ userActivity: NSUserActivity) {
        logger.debug("User Activity Received")
        logger.debug("User Activity Type: \(userActivity.activityType)")
        logger.debug("User Activity Title: \(userActivity.title ?? "")")
        logger.debug("User Activity User Info: \(userActivity.userInfo ?? [:])")

        if userActivity.activityType == NSUserActivityTypeBrowsingWeb {
            logger.debug("Opening UserActivity URL: \(userActivity.webpageURL?.absoluteString ?? "")")
        }

        _ = appNotificationHandler.continueUserActivity(
            userActivity,
            restorationHandler: { _ in }
        )
    }
}
